### PR TITLE
Implement AWS Redshift provider

### PR DESCRIPTION
**Pull request by reflectored**
## Description

This PR implements the AWS Redshift provider, with the support for [datashares](https://docs.aws.amazon.com/redshift/latest/dg/datashare-overview.html) and (initially) the Geometry type.

### CODE REVIEW

File: src/providers/amazon-redshift/qgsredshifttablemodel.cpp [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+      else
+      {
+        item->setFlags( item->flags() & ~Qt::ItemIsSelectable );
+
+        if ( i == DbtmDatabase )
+          item->setData( QgsApplication::getThemeIcon(
+                           QStringLiteral( "/mIconWarning.svg" ) ),
+                         Qt::DecorationRole );
+
+        if ( i == DbtmDatabase || DbtmSchema || i == DbtmTable ||
```
Comment (jef-n) [inline src/providers/amazon-redshift/qgsredshifttablemodel.cpp:431]: i == DbtmSchema

**Comment by jef-n** [comment]
Looks quite near to the postgresql provider - it was forked from a pretty old version.  Could you give a bit of background, why this needs to be separated from the postgres provider?  So far I only found renames, indentation changes, removed service (should predate the fork I think) and legacy support and missing meanwhile added features (query logging, Qt6 updates, metadata, raster support, table filtering, geography, curves, transactions, notify) and just the different table lookup and the additional database level when browsing. But it's hard to tell which things were not there yet, when the provider was forked or actively removed and/or added to support redshift.

**Comment by reflectored** [comment]
Hi Jürgen, thanks for looking into this. As you have noticed, the AWS Redshift provider development has started two years ago and relied on many pieces from the Postgres provider, due to the similarities in the SQL syntax. However there are discrepancies between what is and will be supported by AWS Redshift and PostGIS, due to pending features and/or architectural differences between Postgres and Redshift (Redshift is a data warehouse solution, it distributes computation, whereas Postgres is a database solution, it runs on a single node). Here is the list of differences that required handling the Redshift driver differently:

- The trivial, AWS Redshift is a data warehouse solution whereas Postgres is a database solution, it also has a different logo.
- Querying tables from external databases, syntax that does not exist in Postgres. This requires parsing the URI differently and handling and persisting the external database name.
- The cursor fetch size for Redshift relies on the amount of compute nodes configured for the cluster/Serverless instance. 
- AWS Redshift does not support geometry/geography spatial type column constraints (yet) - the geometry/geography spatial type is not available in the table metadata.
- Parsing [datashare](https://docs.aws.amazon.com/redshift/latest/dg/concepts.html) table info.
- The following Postgres/GIS features are not supported by AWS Redshift:
  - The view spatial_ref_sys.
  - INSERT ... RETURNING syntax.
  - listagg is not supported in sub queries.
- Integration test results may differ in some cases from Postgres/GIS.

**Comment by nyalldawson** [comment]
@reflectored 

Honestly, none of those reasons seem particularly worthy of completely forking the existing provider and adding an additional 17k+ lines of near duplicate code.

I'd much rather see a refactoring of the existing postgres provider to abstract out all the components which aren't common between postgres/redshift vs introducing the duplicate code.

**Comment by reflectored** [comment]
@jef-n I have found the commit the work initially started on 1a83d21912. @nyalldawson I will investigate this option.

**Pull request closed without merging**